Add git info to check/push

JakeDawkins opened the pull request:
Adds the ability to pass git info to all check/push commands for use in Apollo Studio

- Adds same sanitizing rules and test cases from https://github.com/apollographql/apollo-tooling/blob/master/packages/apollo/src/git.ts#L36-L60
- Adds a new `git` module to the crate
- Fetches the git info from the directory that the command is being run in. This assumes it's being run in the project's git directory and not a parent folder. We can get around this with future flags for specific overrides.
- If there is no git info available, that's fine and won't error :)

JakeDawkins committed:
WIP

JakeDawkins committed:
WIP: figure out remote rules

JakeDawkins committed:
feat(git): add local git context to relevant commands

JakeDawkins committed:
Merge branch 'main' into jake/check-git

JakeDawkins committed:
lol linting is fun

JakeDawkins committed:
Merge branch 'main' into jake/check-git

EverlastingBugstopper commented on src/command/graph/push.rs:
```diff
@@ -43,11 +44,21 @@ impl Push {
+                    commit: git.commit,
+                    message: git.message,
+                    remote_url: git.remote_url,
+                })
```
i wonder if @lrlna knows of any special tricks to make this conversion a bit less verbose since all of the field names are the exact same

lrlna replied:
yes, there is a way! Because `GitContextInput` and `GitContext` are two different types, we need to implement an [`Into` trait](https://doc.rust-lang.org/std/convert/trait.Into.html) for `GitContext`. Something like this:

```rust
struct GitContext;
impl Into<push::push_schema_mutation::GitContextInput> for GitContext {
    fn into(self) -> GitContextInput {
        GitContextInput {
          branch: self.branch, // all other fields   
       }
    }
}
```

You will have to do this for `check::` version of `GitContextInput` as well. And then here, instead of doing this iteration you can just go `git.into()`, which is very sleek indeed :). (Fun fact, if they _were_ the same type, we could have just done `...`, but one can only dream, eh)

EverlastingBugstopper replied:
i knew it! this sounds perfect 😄

JakeDawkins committed:
feat(git): impl Into for GitContext`

JakeDawkins committed:
Merge branch 'jake/check-git' of github.com:apollographql/rover into jake/check-git

JakeDawkins committed:
feat(git): move context generation

EverlastingBugstopper commented on src/cli.rs:
```diff
@@ -68,6 +71,20 @@ impl Rover {
+        let committer = self.env_store.get(RoverEnvKey::GitCommitter).unwrap();
+        let commit = self.env_store.get(RoverEnvKey::GitCommit).unwrap();
+        // this url is still dirty, make sure not to use it anywhere else :)
+        let remote_url = self.env_store.get(RoverEnvKey::GitRemoteUrl).unwrap();
```
these should all be `?`

JakeDawkins replied:
Hehe riiiiiiight

JakeDawkins replied:
Oh wait, no. Because I don't want to return from the fn if one of those aren't there

JakeDawkins replied:
But I also don't want panics 🤔 let me fixxxxxx

JakeDawkins replied:
Got it! `unwrap_or` does the trick :)

EverlastingBugstopper commented on src/cli.rs:
```diff
@@ -68,6 +71,20 @@ impl Rover {
+
+        // constructing GitContext with a set of overrides from env vars
+        let git_context = GitContext::new(branch, committer, commit, remote_url);
+        tracing::debug!("Git Context: {:?}", git_context);
```
```suggestion
        tracing::debug!("Git Context", ?git_context);
```

JakeDawkins replied:
This doesn't work for me, I get an error saying 

```
error: expected expression, found `?`
  --> src/cli.rs:97:40
   |
97 |         tracing::debug!("Git Context", ?git_context);
   |                                        ^ expected expression

```

EverlastingBugstopper commented on src/command/graph/check.rs:
```diff
@@ -31,14 +32,19 @@ pub struct Check {
                 graph_id: self.graph.name.clone(),
                 variant: Some(self.graph.variant.clone()),
                 schema: Some(sdl),
+                git_context: Some(git_context.into()),
```
we should either make `git_context` a real `Option` that can sometimes be `None` (if you're not in a git repo for instance) _or_ we should just not make this an `Option` since all the fields _in_ `GitContext` are `Option`s

JakeDawkins replied:
I made the git_context required by the operations, so we can drop all the `Some()`s 💯

EverlastingBugstopper commented on src/git.rs:
```diff
@@ -0,0 +1,300 @@
+            message: self.message,
+        }
+    }
+}
```
really nice to bring this out into this file so it's hidden - i'd love if we could put it at the bottom of the file though since it's just boilerplate

JakeDawkins replied:
I'll move it to right above the tests. I feel like the very bottom would feel weird to me

EverlastingBugstopper commented on src/git.rs:
```diff
@@ -0,0 +1,300 @@
+    ) -> Self {
+        let git = git_info::get();
+
+        let branch = if override_branch.is_none() {
```
these could all be one liners 

```rust
let branch = override_branch.unwrap_or_else(|| git.current_branch);
```

JakeDawkins replied:
This doesn't quite work, because `git.current_branch` is an `Option<String>`, so the return types clash of it either being a `string` from the unwrapping of `override_branch` or an `Option<String>` from git.

lrlna replied:
you're looking for `or()`! So:
```rust
let branch = override_branch.or(git.current_branch);
```

EverlastingBugstopper commented on src/git.rs:
```diff
@@ -0,0 +1,300 @@
+                    if user.is_empty() && email.is_empty() {
+                        None
+                    } else {
+                        Some(format!("{} {}", user, email))
```
this gets a bit murky if user is empty and email is not - wonder if that's a common case

lrlna replied:
definitely a common case i've seen a whole bunch

EverlastingBugstopper commented on src/git.rs:
```diff
@@ -0,0 +1,300 @@
+            GitContext::sanitize_remote_url(&remote)
+        } else {
+            None
+        };
```
```rust
let remote_url = remote_url.map(|url| GitContext::sanitize_remote_url(&url));
```

JakeDawkins replied:
The result of this would also be the wrong type, it'd be `Option<Option<String>>` instead of `Option<String>`

EverlastingBugstopper commented on src/git.rs:
```diff
@@ -0,0 +1,300 @@
+                } else {
+                    remote_url
+                };
+
```
this block is a bit confusing to me - calling `config.remove` returns a value that we then use? is there a `get` method we could use instead?

JakeDawkins replied:
Me and irina were looking at this, and I think this is all we have. It's because we need ownership of the returned value. Everything else we could find returns a borrowed value. I'll add a comment

JakeDawkins committed:
update with review comments

JakeDawkins committed:
clean up unneeded match

JakeDawkins committed:
fix: make git_context non-nullable, since we always have it

JakeDawkins committed:
Merge branch 'main' into jake/check-git

JakeDawkins committed:
fix: linting issues

JakeDawkins committed:
Merge branch 'jake/check-git' of github.com:apollographql/rover into jake/check-git

lrlna committed:
chore(git): refactor GitContext struct

- creates GitContext::new() that takes no arguments
- creates GitContext::with_env() that uses RoverEnv store to create git
object
- separates out GitContext::committer() and GitContext::remote() into
their own logic methods

lrlna committed:
chore: rename git env vars to vcs env vars
